File: apps/theorem-harness-node/smoke.mjs
// JS smoke test for the theorem-harness Node binding.
//
// Proves the round-trip JS -> Rust SDK -> GraphStore -> back to JS against the
// built .node addon. Build first:
//   cargo build --manifest-path apps/theorem-harness-node/Cargo.toml
//   cp apps/theorem-harness-node/target/debug/libtheorem_harness_node.dylib \
//      apps/theorem-harness-node/theorem_harness_node.node
//   node apps/theorem-harness-node/smoke.mjs

import { createRequire } from "module";
import { fileURLToPath } from "url";
import { dirname, join } from "path";

const require = createRequire(import.meta.url);
const here = dirname(fileURLToPath(import.meta.url));
const { Harness } = require(join(here, "theorem_harness_node.node"));

const harness = new Harness();

const runId = harness.startRun("demo from node", "node-smoke", "k-create");
console.log("started run:", runId);

let events = JSON.parse(harness.eventsJson(runId));
console.log("after start:", events.length, events.map((e) => e.kind));

harness.cancel(runId, "stopping from node", "k-cancel");
events = JSON.parse(harness.eventsJson(runId));
console.log("after cancel:", events.length, events.map((e) => e.kind));

const text = harness.pollText(runId, 0);
console.log("text view:", JSON.stringify(text));

const ok =
  events.length === 2 &&
  events[0].kind === "Created" &&
  events[1].kind === "Cancelled" &&
  text.includes("stopping from node");

if (ok) {
  console.log("SMOKE PASS");
} else {
  console.error("SMOKE FAIL");
  process.exit(1);
}

File: apps/theorem-harness-swift/smoke.swift
// Swift smoke for the theorem-harness Swift (UniFFI) binding.
//
// Proves the round-trip Swift -> generated bindings -> Rust SDK -> RedCore -> Swift,
// the same lifecycle the Node smoke exercises, against the SAME Rust core. Build:
//   cargo build
//   cargo run --bin uniffi-bindgen -- generate \
//     --library target/debug/libtheorem_harness_swift.dylib --language swift --out-dir generated
//   swiftc -parse-as-library -emit-executable -o smoke_swift \
//     -L target/debug -ltheorem_harness_swift \
//     -Xcc -fmodule-map-file=generated/theorem_harness_swiftFFI.modulemap -I generated \
//     generated/theorem_harness_swift.swift smoke.swift
//   DYLD_LIBRARY_PATH=target/debug ./smoke_swift

import Foundation

func kinds(_ json: String) -> [String] {
    guard let data = json.data(using: .utf8),
        let arr = (try? JSONSerialization.jsonObject(with: data)) as? [[String: Any]]
    else { return [] }
    return arr.map { ($0["kind"] as? String) ?? "?" }
}

@main
struct Smoke {
    static func main() {
        let dataDir = (NSTemporaryDirectory() as NSString)
            .appendingPathComponent("theorem-harness-swift-" + UUID().uuidString)
        try? FileManager.default.createDirectory(atPath: dataDir, withIntermediateDirectories: true)
        print("data dir:", dataDir)

        do {
            let harness = try Harness(dataDir: dataDir)

            let runId = try harness.startRun(
                task: "demo from swift", actor: "swift-smoke", idempotencyKey: "k-create")
            print("started run:", runId)

            let afterStart = kinds(try harness.eventsJson(runId: runId))
            print("after start:", afterStart)

            try harness.cancel(
                runId: runId, reason: "stopping from swift", idempotencyKey: "k-cancel")
            let afterCancel = kinds(try harness.eventsJson(runId: runId))
            print("after cancel:", afterCancel)

            let status = try harness.runStatus(runId: runId)
            print("status:", status)

            _ = try harness.remember(
                agentId: "swift-smoke", kind: "belief", title: "binding is durable",
                content: "The Swift binding persists to RedCore.")
            let recalledJson = try harness.recall(agentId: "swift-smoke", query: "binding", limit: 10)
            let recalledCount =
                ((try? JSONSerialization.jsonObject(with: recalledJson.data(using: .utf8)!)) as? [Any])?
                .count ?? 0
            print("recalled:", recalledCount)

            let ok =
                afterStart == ["Created"] && afterCancel == ["Created", "Cancelled"]
                && status == "cancelled" && recalledCount >= 1
            print(ok ? "SMOKE PASS" : "SMOKE FAIL")
            exit(ok ? 0 : 1)
        } catch {
            print("SMOKE FAIL:", error)
            exit(1)
        }
    }
}
